src/translate/axiom_rules.py

```diff
@@ -257,18 +244,20 @@ def compute_axiom_layers(clusters, strategy):
 
 def compute_negative_axioms(clusters):
     for cluster in clusters:
-        if cluster.needed_negative:
+        if cluster.needed_negatively:
             if len(cluster.variables) > 1:
-                # If the cluster contains several variables we need to
-                # overapproximate when a derived predicate can become negative.
-                # We do this by simply allowing it to always become negative
-                # unconditionally.
-                # The successor generator in the search component does not
-                # consider these overapproximations since they set the derived
-                # variable to its default value (false).
-                # The reason for adding the overapproximation at all is that
-                # some heuristics (e.g. relaxation and causal graph heuristics)
-                # use them to derive that the negated value is reachable.
+                # If the cluster contains several variables they depend
+                # (positively) on each other and the normal negation computation
+                # can lead to some heuristics incorrectly detecting states as
+                # unsolvable because they don't consider negation by failure and
+                # conclude that a derived variable cannot be reached negatively
+                # (for details see issue453).
+                # We avoid this by instead giving a naive overapproximation
+                # saying that the negated variable can always be reached with
+                # no conditions.
+                # This does not interfere with the sucessor generator in the
+                # search component since it only considers axioms which set a
+                # derived variable to its non-default value.
```

Comment:
I think the first sentence misses some commas, and I'm not sure how clear the discussion about heuristics is. Regarding the last sentence, I don't see a connection to the successor generator? The successor generator is the part that looks at which operators are applicable, it's not directly related to axioms.

I would suggest something like (rewrapping, but keeping the paragraph breaks):

```
If the cluster contains multiple variables, they have a cyclic positive dependency. In this case, the "obvious" way of negating the formula defining the derived variable is semantically wrong. For details, see issue453.

Therefore, in this case we perform a naive overapproximation instead, which assumes that derived variables occurring in such clusters can be false unconditionally. This is good enough for correctness of the code that uses these negated axioms (within heuristics of the search component), but loses accuracy. Negating the rules in an exact (non-overapproximating) way is possible but more expensive. Again, see issue453 for details.

TODO: It would be cleaner if these negated rules were an implementation detail of the heuristics in the search component that make use of them rather than part of the translation process. They should be removed in the future. Similarly, it would be a good idea to remove the notion of axiom layers and derived variable default values from the output. (All derived variables should be binary and default to false.)
```

The last sentence doesn't really fit here and would perhaps make more sense as a TODO somewhere else in the file. But I think it would be good to include it somewhere to not lose these aspects of the discussions we had for this issue.